paperqa/llms.py

```diff
@@ -764,6 +802,378 @@ async def select_tool(
+
+        with io.BytesIO() as tmp_file:
+            await self.write_jsonl(messages, tmp_file)
+            tmp_file.seek(0)
+            file = await client.files.create(file=tmp_file, purpose="batch")
+
+        batch = await client.batches.create(
+            input_file_id=file.id,
+            endpoint="/v1/chat/completions",
+            completion_window="24h",
+            metadata={"description": ""},
+        )
+
+        start_clock = asyncio.get_running_loop().time()
+        while batch.status != BatchStatus.COMPLETE.from_openai():
+            batch = await client.batches.retrieve(batch.id)
+            if batch.status == BatchStatus.FAILURE.from_openai():
+                error_messages = []
+                if batch.errors and hasattr(batch.errors, "data") and batch.errors.data:
+                    error_messages = [
+                        str(k.message)
+                        for k in batch.errors.data
+                        if k.message is not None
+                    ]
+                raise RuntimeError(
+                    "Batch failed. \n\nReason: \n" + "\n".join(error_messages)
+                )
+            if batch.status == BatchStatus.CANCEL.from_openai():
+                raise ConnectionError("Batch was cancelled.")
+
+            batch_time = asyncio.get_running_loop().time() - start_clock
+            if batch_time > self.config.get("batch_summary_time_limit", 24 * 60 * 60):
+                raise TimeoutError("Batch took too long to complete.")
+
+            logger.info(
+                f"Summary batch status: {batch.status} | Time elapsed: {batch_time}"
+            )
+            await asyncio.sleep(self.config.get("batch_polling_interval", 30))
+
+        if batch.output_file_id:
+            api_responses = await client.files.content(batch.output_file_id)
+        else:
+            raise RuntimeError("Batch failed to generate output file.")
+        sorted_responses = sorted(
+            [
+                json.loads(line)
+                for line in api_responses.read().decode("utf-8").splitlines()
+            ],
+            key=lambda x: int(x["custom_id"]),
+        )  # The batchAPI doesn't guarantee the order of the responses
+
+        return [
+            Chunk(
+                text=response["response"]["body"]["choices"][0]["message"]["content"],
+                prompt_tokens=response["response"]["body"]["usage"]["prompt_tokens"],
+                completion_tokens=response["response"]["body"]["usage"][
+                    "completion_tokens"
+                ],
+            )
+            for response in sorted_responses
+        ]
+
+    async def achat_iter(self):
+        raise NotImplementedError(
+            "Async generator not supported for batch calls. Use achat instead."
+        )
+
+
+class AnthropicBatchLLMModel(LLMBatchModel):
+    """A wrapper around the anthropic library to use the batch API."""
+
+    name: str = "claude-3-5-sonnet-20241022"
```

Comment:
Can you add a `description` why we are using an exact name here? Imo we should prolly just use `claude-3-5-sonnet-latest ` (and not specify a date)